gpytorch/lazy/lazy_tensor.py

```diff
@@ -1590,9 +1583,30 @@ def diagonalization(self, method: Optional[str] = None):
 
         elif method == "symeig":
             evals, evecs = self.symeig(eigenvectors=True)
+        else:
+            raise RuntimeError(f"Unknown diagonalization method '{method}'")
 
         return evals, evecs
 
+    def _choose_root_method(self, method: Optional[str] = None) -> str:
+        # Better inform which root_decomposition or root_inv_decomposition
+        # method to use based on available caches and matrix size.
+        if _is_in_cache_ignore_all_args(self, "symeig"):
+            return "symeig"
+        if _is_in_cache_ignore_all_args(self, "diagonalization"):
+            return "diagonalization"
+        if _is_in_cache_ignore_all_args(self, "lanczos"):
+            return "lanczos"
```

Comment:
I like how you pulled out this functionality to keep things tidy. 

One concern is that this may introduce a bit of a gotcha, since it's not really transparent to the user that the `method` selection will be overridden silently (I guess that was the case prior to this change as well, though). That is, say you compute the root via Lanczos first, and are dissatisfied with the quality of the approximation and thus want to use `symeig` instead. If you just call `root_decomposition` with the `symeig` method you just going to get lanczos back from the cache without knowing.

One thing we could do is add a `force_method` arg to this method (and to `root_decomposition`) that used the desired method regardless of what's in cache. An alternative would be to issue a warning if the desired method does not end up getting selected due to other roots being readily available in cache. 

Curious to hear what @jacobrgardner, @gpleiss, @wjmaddox think about this.

Reply:
At least from my experience, this logic is a pretty strong gotcha.  Returning lanczos decompositions when you're intending to specify `cholesky` can result in some pretty sub-optimal behavior. I think there's some downstream code (SKI fantasy strategy) that could end up having both lanczos and cholesky roots being computed at different times, and there I know that the cholesky is definitely required at a certain time. Maybe just wrapping this code in a if method is None tag will clean it up in the default setting?

Reply:
It's an interesting point because
* In the case of a 1x1 matrix we're ignoring method and just doing 1/sqrt(self), and I think we want to keep that behavior
* The existing implementation actually checked the `diagonalization` cache before some of the methods, so we'll be changing that behavior
I agree we should use largely use method when specified though. I'll change the logic to do that.

Reply:
> In the case of a 1x1 matrix we're ignoring method and just doing 1/sqrt(self), and I think we want to keep that behavior

Definitely. I don't think we need to do lanczos to take square roots. We could add a `babylonian` method maybe :)

Reply:
So are we good to go with this PR?